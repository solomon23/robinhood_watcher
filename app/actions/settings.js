// @flow
import { getSettings } from '../services/api'
import * as ipc from '../services/ipc'

export const VIEW_BY = {
  PERCENT: 'PERCENT',
  TOTAL: 'TOTAL',
  INDIVIDUAL: 'INDIVIDUAL',
}

export const SETTINGS_UPDATE = 'SETTINGS_UPDATE'
export function updateSettings(settings: UserSettings) {
  return { type: SETTINGS_UPDATE, payload: settings }
}

export function loadSettings() {
  return (dispatch: Dispatch) => {
    // get the settings from the local store
    const settings = getSettings()

    if (settings) {
      dispatch(updateSettings(settings))
    }
  }
}

export function saveSettings(settings: UserSettings) {
  return (dispatch: Dispatch) => {
    // update it locally
    dispatch(updateSettings(settings))

    // save it over the wire
    ipc.saveSettings(settings)
  }
}

export function quit() {
  return () => {
    ipc.quit()
  }
}

export function logout() {
  return () => {
    ipc.logout()
  }
}
