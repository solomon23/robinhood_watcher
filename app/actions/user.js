import { authenticate, tryAuthenticate } from '../services/api'
import type { Dispatch } from '../reducers/types'

export const LOGIN_USER_LOADING = 'LOGIN_USER_LOADING'
export const LOGIN_USER_ERROR = 'LOGIN_USER_ERROR'
export const LOGIN_USER_SUCCESS = 'LOGIN_USER_SUCCESS'
export const LOGIN_USER_NEEDS_MFA = 'LOGIN_USER_NEEDS_MFA'
export const USER_LOGOUT = 'USER_LOGOUT'
export const SET_USER_ACCOUNT_NUMBER = 'SET_USER_ACCOUNT_NUMBER'

export function login(username, password, mfaCode) {
  return (dispatch: Dispatch) => {
    dispatch({ type: LOGIN_USER_LOADING })

    return authenticate(username, password, mfaCode)
      .then(response => {
        if (response.twoFactorAuthRequired && !mfaCode) {
          dispatch({ type: LOGIN_USER_NEEDS_MFA })
        } else if (response.authenticated) {
          dispatch({ type: LOGIN_USER_SUCCESS, response })
        } else {
          dispatch({ type: LOGIN_USER_ERROR, message: response.error })
        }

        return null
      }).catch(err => {
        dispatch({ type: LOGIN_USER_ERROR, error: err })
      })
    }
}

export function tryAuth() {
  return (dispatch: Dispatch) => {
    dispatch({ type: LOGIN_USER_LOADING })

    return tryAuthenticate()
      .then(response => {
        if (response.authenticated) {
          dispatch({ type: LOGIN_USER_SUCCESS, response })
        } else {
          dispatch({ type: LOGIN_USER_ERROR, message: response.error })
        }

        return null
      }).catch(err => {
        dispatch({ type: LOGIN_USER_ERROR, error: err })
      })
    }
}

export function logout() {
  return {
    type: USER_LOGOUT,
  }
}