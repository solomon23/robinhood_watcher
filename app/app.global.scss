/*
 * @NOTE: Prepend a `~` to css file paths that are in your node_modules
 *        See https://github.com/webpack-contrib/sass-loader#imports
 */
@import "~@fortawesome/fontawesome-free/css/all.css";

$green: #21ce99;
$red: #f45531;

body {
  position: relative;
  color: white;
  height: 100vh;
  background-color: #212025;
  font-family: Arial, Helvetica, Helvetica Neue, serif;
  overflow-y: hidden;
  margin: 0;
}
