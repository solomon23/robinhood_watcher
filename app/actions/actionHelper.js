// @flow
import * as userActions from './user'
import * as appActions from './app'

export function createRequestTypes(base: string) {
  return ['REQUEST', 'SUCCESS', 'FAILURE'].reduce((acc, type) => {
    acc[type] = `${base}_${type}`
    return acc
  }, {})
}

export function action(type: string, payload: any = {}) {
  return { type, ...payload }
}

export function createRequests(base: any) {
  return {
    request: (o: any) => action(base.REQUEST, o),
    success: (o: any) => action(base.SUCCESS, o),
    failure: (o: any) => action(base.FAILURE, o),
  }
}

export function doApiCall(options: any) {
  const defaults = {
    api: () => { },
    action: {},
    apiParams: {},
    guardFunc: () => true,
    postFunc: (dispatch, getState, response) => response,
  }
  const opts = { ...defaults, ...options }
  opts.action = Array.isArray(options.action) ? options.action : [options.action]

  return (dispatch: Dispatch, getState: GetState) => {
    // bail if we don't pass any rules sent in
    if (!opts.guardFunc(getState())) {
      return Promise.resolve()
    }

    const { user } = getState()
    if (!user.authenticated) {
      // bail
      return Promise.resolve()
    }

    const params = { ...opts.apiParams || {}, user }

    // mark the call as starting
    opts.action.map(act => dispatch(act.request({ request: params })))

    // call the api with the args
    return opts.api(params)
      .then(
        response => opts.postFunc(dispatch, getState, response)
      ).then((response) => {
        opts.action.map(act => dispatch(act.success({ request: params, response })))

        return response
      }, (error) => {
        opts.action.map(act => dispatch(act.failure({ request: params, response: error })))
        return Promise.reject(error)
      }).catch((error) => {
        if (error === 401) {
          // log the user out and stop refreshing the data since our auth token is no longer valid
          dispatch(userActions.logout())
          dispatch(appActions.stopRefresh())
        }
      })
  }
}
