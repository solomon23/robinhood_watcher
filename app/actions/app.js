// @flow
import * as API from '../services/api'
import { createRequests, doApiCall, createRequestTypes } from './actionHelper'
import { createSettingsWindow } from '../services/windows'

export const START_REFRESH_TIMER = 'START_REFRESH_TIMER'
export const STOP_REFRESH_TIMER = 'STOP_REFRESH_TIMER'
export const APP_QUIT = 'APP_QUIT'

export const GET_ACCOUNT_NUMBER = createRequestTypes('GET_ACCOUNT_NUMBER')
export const getAccountNumberActions = createRequests(GET_ACCOUNT_NUMBER)

export const GET_PORTFOLIO = createRequestTypes('GET_PORTFOLIO')
export const getPortfolioActions = createRequests(GET_PORTFOLIO)

export const GET_ACCOUNT_POSITIONS = createRequestTypes('GET_ACCOUNT_POSITIONS')
export const getAccountPositionsActions = createRequests(GET_ACCOUNT_POSITIONS)

export const GET_WATCH_LIST = createRequestTypes('GET_WATCH_LIST')
export const getWatchListActions = createRequests(GET_WATCH_LIST)

export function startRefreshTimer() {
  return { type: START_REFRESH_TIMER }
}

export function stopRefreshTimer() {
  return { type: STOP_REFRESH_TIMER }
}

export function appQuit() {
  return { type: APP_QUIT }
}

export function getAccount() {
  return doApiCall({
    api: API.getAccountNumber,
    action: getAccountNumberActions,
    guardFunc: (state) => {
      const { user } = state
      return !user.accountNumber
    },
  })
}

export function getPortfolio() {
  return doApiCall({
    api: API.getPortfolio,
    action: getPortfolioActions,
  })
}

export function getAccountPositions() {
  return doApiCall({
    api: API.getAccountPositions,
    action: getAccountPositionsActions,
  })
}

export function getWatchList() {
  return doApiCall({
    api: API.getWatchlist,
    action: getWatchListActions,
  })
}

export function getAllData() {
  return async function getAllCallback(dispatch: Dispatch, getState: GetState) {
    const { user } = getState()

    // bail if the user isn't logged in
    if (!user.authenticated) {
      return
    }

    if (!user.accountNumber) {
      await getAccount()(dispatch, getState)
    }

    await getPortfolio()(dispatch, getState)
    await getAccountPositions()(dispatch, getState)
    await getWatchList()(dispatch, getState)
  }
}

export function openSettings(settings: UserSettings) {
  return () => {
    createSettingsWindow(settings)
  }
}
