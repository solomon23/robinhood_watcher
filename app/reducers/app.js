// @flow
import * as appActions from '../actions/app'
import type { Action } from './types'

const defaultState = {
  actions: {
    portfolioLoading: false,
    positionsLoading: false,
    refreshing: false,
  },
  portfolio: {
  },
  positions: [
  ],
}

export default function app(state = defaultState, action: Action) {
  switch (action.type) {
    case appActions.START_REFRESH_TIMER: {
      const actions = { ...state.actions, refreshing: true }
      return { ...state, actions }
    }

    case appActions.STOP_REFRESH_TIMER: {
      const actions = { ...state.actions, refreshing: false }
      return { ...state, actions }
    }

    case appActions.GET_PORTFOLIO.REQUEST: {
      const actions = { ...state.actions, portfolioLoading: true }
      return { ...state, actions }
    }

    case appActions.GET_PORTFOLIO.FAILURE: {
      const actions = { ...state.actions, portfolioLoading: false }
      return { ...state, actions }
    }

    case appActions.GET_PORTFOLIO.SUCCESS: {
      const actions = { ...state.actions, portfolioLoading: false }
      return { ...state, actions, portfolio: action.response }
    }

    case appActions.GET_ACCOUNT_POSITIONS.REQUEST: {
      const actions = { ...state.actions, positionsLoading: true }
      return { ...state, actions }
    }

    case appActions.GET_ACCOUNT_POSITIONS.FAILURE: {
      const actions = { ...state.actions, positionsLoading: false }
      return { ...state, actions }
    }

    case appActions.GET_ACCOUNT_POSITIONS.SUCCESS: {
      const actions = { ...state.actions, positionsLoading: false }
      return { ...state, actions, positions: action.response }
    }

    default:
      return state
  }
}
