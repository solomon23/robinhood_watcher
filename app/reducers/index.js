// @flow
import { combineReducers } from 'redux'
import { connectRouter } from 'connected-react-router'
import toolbar from './toolbar'
import user from './user'
import app from './app'
import settings from './settings'

export default function createRootReducer(history: {}) {
  const routerReducer = connectRouter(history)(() => {})

  return connectRouter(history)(
    combineReducers({
      router: routerReducer,
      toolbar,
      app,
      user,
      settings,
    })
  )
}
