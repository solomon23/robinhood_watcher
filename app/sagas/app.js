import { takeEvery, fork, put, call, race, take } from 'redux-saga/effects'
import { ipcRenderer } from 'electron'
import * as appActions from '../actions/app'
import * as userActions from '../actions/user'
import * as api from '../services/api'

const REFRESH_TIME = 1000 * 60 * 10

const wait = ms => (
  new Promise(resolve => {
    setTimeout(() => resolve(), ms)
  })
)

function* handleStartRefresh() {
  yield takeEvery(appActions.START_REFRESH_TIMER, function* handler() {
    // fire it off first
    yield put(appActions.getAllData())

    while(true) {
      const timer = yield race({
        stopped: take(appActions.STOP_REFRESH_TIMER),
        tick: call(wait, REFRESH_TIME),
      })

      if (!timer.stopped) {
        yield put(appActions.getAllData())
      } else {
        break
      }
    }
  })
}

function* handleLogout() {
  yield takeEvery(userActions.USER_LOGOUT, function* handler() {
    // stop refreshing
    yield put (appActions.stopRefresh())

    // kill our cookie
    api.logout()
  })
}

function* appQuit() {
  yield takeEvery(appActions.APP_QUIT, () => {
    // send the quit
    ipcRenderer.send('APP_QUIT')
  })
}

function* app() {
  yield fork(handleStartRefresh)
  yield fork(handleLogout)
  yield fork(appQuit)
}

export default app
