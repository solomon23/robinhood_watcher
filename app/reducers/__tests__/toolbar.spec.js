import reducer from '../toolbar'
import * as ActionTypes from '../../actions/app'

describe('toolbar reducer', () => {
  it('should handle GET_PORTFOLIO.SUCCESS', () => {
    expect(
      reducer({}, {
        type: ActionTypes.GET_PORTFOLIO.SUCCESS,
        response: {
          equity: 100,
        },
      })
    ).toMatchSnapshot()
  })

  it('should handle GET_PORTFOLIO.SUCCESS after hours', () => {
    expect(
      reducer({}, {
        type: ActionTypes.GET_PORTFOLIO.SUCCESS,
        response: {
          equity: 10,
          extended_hours_equity: 200,
        },
      })
    ).toMatchSnapshot()
  })
})
