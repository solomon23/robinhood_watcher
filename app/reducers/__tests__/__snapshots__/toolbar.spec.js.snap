// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toolbar reducer should handle GET_PORTFOLIO.SUCCESS 1`] = `
Object {
  "title": "$100.00",
}
`;

exports[`toolbar reducer should handle GET_PORTFOLIO.SUCCESS after hours 1`] = `
Object {
  "title": "$200.00",
}
`;
